flows/common/common.go

```diff
@@ -68,36 +80,84 @@ func Handler(fn hdrsFunc, match matchFunc, reporter *Reporter) lwotg.FlowGenerat
 			size := len(buf.Bytes())
 
 			klog.Infof("%s Tx interface %s", flow.Name, tx)
-			handle, err := pcap.OpenLive(tx, 1500, true, pcapTimeout)
+
+			ih, err := pcap.NewInactiveHandle(tx)
+			if err != nil {
+				klog.Errorf("cannot create handle, err: %v", err)
+				return
+			}
+			defer ih.CleanUp()
+			if err := ih.SetImmediateMode(true); err != nil {
+				klog.Errorf("cannot set immediate mode on handle, err: %v", err)
+				return
+			}
+
+			if err := ih.SetPromisc(false); err != nil {
+				klog.Errorf("cannot set promiscous mode, err: %v", err)
+				return
+			}
+
+			if err := ih.SetSnapLen(packetBytes); err != nil {
+				klog.Errorf("cannot set packet length, err: %v", err)
+			}
+
+			// Set buffer to be 200MB of sent packets.
+			if err := ih.SetBufferSize(200e6); err != nil {
+				klog.Errorf("cannot set buffer size, err: %v", err)
+			}
+
+			handle, err := ih.Activate()
 			if err != nil {
 				klog.Errorf("%s Tx error: %v", flow.Name, err)
 				return
 			}
 			defer handle.Close()
 
 			f.setTransmit(true)
+			totPackets := uint32(0)
+			// stopFlow indicates whether we should stop sending packets on the flow, it is set
+			// when the flow specification says that we should only send a limited number of
+			// packets.
+			var stopFlow bool
 			for {
 				select {
 				case <-stop:
 					klog.Infof("%s send exiting on %s", flow.Name, tx)
 					f.setTransmit(false)
 					return
 				default:
-					klog.Infof("%s sending %d packets", flow.Name, pps)
-					for i := 1; i <= int(pps); i++ {
-						if err := handle.WritePacketData(buf.Bytes()); err != nil {
-							klog.Errorf("%s cannot write packet on interface %s, %v", flow.Name, tx, err)
-							return
+					switch stopFlow {
+					case true:
+						// avoid busy looping.
```

Comment:
why do we need this busy loop here, as opposed to, say, writing to the stop channel and letting the stop case of the select return?

Reply:
We want to have the telemetry of the flow continue to be updated with the transmit state set to true (i.e., we're transmitting according to the spec and not stopped). This means we need to remain in the same state until we get something on the `stop` channel, hence remaining within the loop.

We could exit and then wait for stop, but this seemed reasonable. We can refactor to break the loop and have a blocking read from `stop`, it seems that this might be more bug prone because if something else breaks without returning we might end up with a blocking read that never returns.